cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -389,9 +399,23 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
   @AuditPolicy(AuditDetail.REQUEST_BODY)
   public void upgradeApplication(HttpRequest request, HttpResponder responder,
                                  @PathParam("namespace-id") String namespaceId,
-                                 @PathParam("app-id") String appName) throws Exception {
-    ApplicationId appId = validateApplicationId(namespaceId, appName);
-    applicationLifecycleService.upgradeApplication(appId, createProgramTerminator());
+                                 @PathParam("app-id") String appName,
+                                 @Nullable @QueryParam("artifactScopes") String artifactScopes,
+                                 @QueryParam("allowSnapshot") @DefaultValue("false") boolean allowSnapshot)
+    throws Exception {
+    ApplicationId appId = validateApplicationId(validateNamespace(namespaceId), appName);
+    Set<ArtifactScope> artifactScopesToConsider = new HashSet<>();
+    if (artifactScopes != null && !artifactScopes.isEmpty()) {
+      for (String scope : Splitter.on(',').trimResults().split(artifactScopes)) {
```

Comment:
instead of having comma separate string, it's better be an array. User can all it like this:

`?artifactScope=USER&artifactScope=SYSTEM`

```
@QueryParam("artifactScope") Set<String> scopes
```